File: sdks/web/tests/client.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { createEventPump, type EventPump } from '../src/client';

const active: EventPump[] = [];

function newClient(): EventPump {
  const client = createEventPump();
  active.push(client);
  return client;
}

interface Captured {
  url: string;
  body: Record<string, any>;
  resolve: (ok?: boolean) => void;
}

let calls: Captured[] = [];
let autoResolve = true;

function fetchMock(url: string | URL, options?: RequestInit): Promise<Response> {
  return new Promise((resolvePromise) => {
    const captured: Captured = {
      url: String(url),
      body: options?.body ? JSON.parse(String(options.body)) : {},
      resolve: (ok = true) =>
        resolvePromise(
          new Response(ok ? '{"accepted":1,"rejected":[]}' : '{"error":"x"}', {
            status: ok ? 200 : 500,
          }),
        ),
    };
    calls.push(captured);
    if (autoResolve) captured.resolve();
  });
}

function identityCalls(): Captured[] {
  return calls.filter((c) => c.url.includes('/v1/identity'));
}

function eventCalls(): Captured[] {
  return calls.filter((c) => c.url.includes('/v1/events'));
}

async function settle(): Promise<void> {
  for (let i = 0; i < 10; i++) await Promise.resolve();
}

const CONFIG = { endpoint: 'https://collect.test', appToken: 'tok-web' };

function clearCookies(): void {
  for (const pair of document.cookie.split(';')) {
    const name = pair.split('=')[0]?.trim();
    if (name) document.cookie = `${name}=; expires=Thu, 01 Jan 1970 00:00:00 GMT`;
  }
}

beforeEach(() => {
  vi.useFakeTimers({ now: 1_700_000_000_000 });
  localStorage.clear();
  sessionStorage.clear();
  clearCookies();
  calls = [];
  autoResolve = true;
  vi.stubGlobal('fetch', fetchMock);
  Object.defineProperty(document, 'visibilityState', { value: 'visible', configurable: true });
});

afterEach(() => {
  active.splice(0).forEach((client) => client.destroy());
  vi.useRealTimers();
  vi.unstubAllGlobals();
});

describe('S0–S4 ordering (SPEC §3)', () => {
  it('registers identity before any events are sent, and buffered events flush after S3', async () => {
    autoResolve = false;
    const ep = newClient();
    ep.init(CONFIG);
    ep.track('product_viewed', { sku: 'A1' });
    await settle();

    expect(identityCalls()).toHaveLength(1);
    expect(eventCalls()).toHaveLength(0); // never sent before S3 completes

    identityCalls()[0]!.resolve();
    await settle();
    ep.flush();
    await settle();

    expect(eventCalls()).toHaveLength(1);
    const sent = eventCalls()[0]!.body.events;
    expect(sent).toHaveLength(1);
    expect(sent[0].event_name).toBe('product_viewed');
    expect(sent[0].properties).toEqual({ sku: 'A1' });
    // identity registration carried the same ids the event uses
    const identity = identityCalls()[0]!.body;
    expect(sent[0].anonymous_id).toBe(identity.anonymous_id);
    expect(sent[0].session_key).toBe(identity.session_key);
    expect(identity.session_number).toBe(1);
    expect(identity.handles.amplitude_device_id).toBe(identity.anonymous_id);
  });

  it('uses the ep_aid cookie and never writes cookies itself', async () => {
    document.cookie = 'ep_aid=0f2937de-92f9-4b6c-a222-abcdefabcdef';
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    expect(identityCalls()[0]!.body.anonymous_id).toBe('0f2937de-92f9-4b6c-a222-abcdefabcdef');
    expect(document.cookie).toBe('ep_aid=0f2937de-92f9-4b6c-a222-abcdefabcdef');
  });
});

describe('session rotation (SPEC §3)', () => {
  it('rotates on visible return after 30 minutes and re-registers', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    const firstKey = identityCalls()[0]!.body.session_key;

    vi.advanceTimersByTime(31 * 60_000);
    document.dispatchEvent(new Event('visibilitychange')); // visible
    await settle();

    expect(identityCalls()).toHaveLength(2);
    const second = identityCalls()[1]!.body;
    expect(second.session_key).not.toBe(firstKey);
    expect(second.session_number).toBe(2);
    expect(second.anonymous_id).toBe(identityCalls()[0]!.body.anonymous_id);
  });

  it('does not rotate within the window', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    ep.track('product_viewed'); // touches last_active_at
    vi.advanceTimersByTime(5 * 60_000);
    document.dispatchEvent(new Event('visibilitychange'));
    await settle();
    expect(identityCalls()).toHaveLength(1);
  });
});

describe('engagement time (SPEC §4)', () => {
  it('each event carries time accumulated since the previous event', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();

    vi.advanceTimersByTime(1_200);
    ep.track('product_viewed');
    vi.advanceTimersByTime(3_400);
    ep.track('product_viewed');
    ep.flush();
    await settle();

    const events = eventCalls().flatMap((c) => c.body.events);
    expect(events[0].context.engagement_time_msec).toBe(1_200);
    expect(events[1].context.engagement_time_msec).toBe(3_400);
    expect(events[0].context.session_number).toBe(1);
    expect(events[0].context.sdk.name).toBe('event-pump-web');
  });
});

describe('user transitions (SPEC §3)', () => {
  it('setUser re-registers with user_id and NEVER rotates anonymous_id', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    const before = identityCalls()[0]!.body;

    ep.setUser('u-42');
    await settle();

    const after = identityCalls()[1]!.body;
    expect(after.user_id).toBe('u-42');
    expect(after.anonymous_id).toBe(before.anonymous_id);
    expect(after.session_key).toBe(before.session_key);

    ep.track('product_viewed');
    ep.flush();
    await settle();
    expect(eventCalls().at(-1)!.body.events[0].user_id).toBe('u-42');
  });

  it('clearUser drops user_id and rotates the session only', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    ep.setUser('u-42');
    await settle();
    const before = identityCalls().at(-1)!.body;

    ep.clearUser();
    await settle();

    const after = identityCalls().at(-1)!.body;
    expect(after.user_id).toBeUndefined();
    expect(after.session_key).not.toBe(before.session_key);
    expect(after.session_number).toBe(2);
    expect(after.anonymous_id).toBe(before.anonymous_id);
  });
});

describe('flush triggers (SPEC §7)', () => {
  it('flushes automatically at 20 buffered events', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    for (let i = 0; i < 20; i++) ep.track('product_viewed');
    await settle();
    expect(eventCalls().flatMap((c) => c.body.events)).toHaveLength(20);
  });

  it('flushes on the 30s timer', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    ep.track('product_viewed');
    await vi.advanceTimersByTimeAsync(30_000);
    await settle();
    expect(eventCalls().flatMap((c) => c.body.events)).toHaveLength(1);
  });

  it('uses sendBeacon with a token query param on hidden', async () => {
    const beacon = vi.fn(() => true);
    Object.defineProperty(navigator, 'sendBeacon', { value: beacon, configurable: true });
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    ep.track('product_viewed');

    Object.defineProperty(document, 'visibilityState', { value: 'hidden', configurable: true });
    document.dispatchEvent(new Event('visibilitychange'));

    expect(beacon).toHaveBeenCalledOnce();
    const [url, payload] = beacon.mock.calls[0] as unknown as [string, string];
    expect(url).toBe('https://collect.test/v1/events?token=tok-web');
    expect(JSON.parse(payload).events).toHaveLength(1);
  });

  it('gives up on events older than 24h', async () => {
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    autoResolve = false;
    ep.track('product_viewed');
    ep.flush();
    await settle();
    calls.at(-1)!.resolve(false); // send fails; stays queued
    await settle();

    vi.advanceTimersByTime(25 * 60 * 60 * 1000);
    autoResolve = true;
    calls = [];
    ep.flush();
    await settle();
    expect(eventCalls()).toHaveLength(0); // dropped, not sent
  });
});

describe('eventHeaders (SPEC §8)', () => {
  it('returns the tier-2 headers', async () => {
    document.cookie = 'ep_aid=0f2937de-92f9-4b6c-a222-abcdefabcdef';
    const ep = newClient();
    ep.init(CONFIG);
    await settle();
    const headers = ep.eventHeaders('add_to_cart');
    expect(headers['X-Event']).toBe('add_to_cart');
    expect(headers['X-Anonymous-Id']).toBe('0f2937de-92f9-4b6c-a222-abcdefabcdef');
    expect(headers['X-Session-Key']).toMatch(/^[0-9a-f-]{36}$/);
  });
});

File: sdks/flutter/test/client_test.dart
import 'dart:ui' show AppLifecycleState;

import 'package:event_pump/event_pump.dart' show sdkName;
import 'package:fake_async/fake_async.dart';
import 'package:flutter_test/flutter_test.dart';

import 'fakes.dart';

void main() {
  late Harness harness;

  tearDown(() => harness.cleanup());

  group('S0–S4 ordering (SPEC §3)', () {
    test('identity registers before any events; buffered events flush after S3', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        harness.transport.manual = true;
        final client = harness.build();

        client.init();
        client.track('product_viewed', properties: {'sku': 'A1'});
        async.flushMicrotasks();

        expect(harness.transport.calls.map((c) => c.$1), ['/v1/identity']);
        expect(harness.transport.sentEvents(), isEmpty);

        harness.transport.completeNext(); // S3 completes -> S4 opens
        async.flushMicrotasks();
        harness.transport.manual = false;
        client.flush();
        async.flushMicrotasks();

        final events = harness.transport.sentEvents();
        expect(events, hasLength(1));
        expect(events.first['event_name'], 'product_viewed');
        expect(events.first['properties'], {'sku': 'A1'});
        final identity = harness.transport.identityBodies().first;
        expect(events.first['anonymous_id'], identity['anonymous_id']);
        expect(events.first['session_key'], identity['session_key']);
        expect(identity['session_number'], 1);
        expect((identity['handles']! as Map)['amplitude_device_id'], identity['anonymous_id']);
        client.dispose();
      });
    });
  });

  group('session persistence across restarts (SPEC §3, approved decision)', () {
    test('cold start within 30 minutes resumes; later cold start rotates', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      final firstKey = first.eventHeaders()['X-Session-Key'];
      first.dispose();

      final soon = harness.build(now: () => base.add(const Duration(minutes: 10)));
      soon.init();
      expect(soon.eventHeaders()['X-Session-Key'], firstKey);
      expect(harness.store.values['ep_session_number'], '1');
      soon.dispose();

      final late = harness.build(now: () => base.add(const Duration(minutes: 45)));
      late.init();
      expect(late.eventHeaders()['X-Session-Key'], isNot(firstKey));
      expect(harness.store.values['ep_session_number'], '2');
      // anonymous_id never rotates with sessions
      expect(late.eventHeaders()['X-Anonymous-Id'], first.eventHeaders()['X-Anonymous-Id']);
      late.dispose();
    });

    test('resuming a live session with a wiped counter posts 1, never 0', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      final firstKey = first.eventHeaders()['X-Session-Key'];
      final firstSeen = harness.store.values['ep_first_seen_at'];
      first.dispose();

      // The counter is gone, the session and the device are not. SPEC §2
      // numbers sessions from 1 and IdentityValidation rejects anything below.
      harness.store.values.remove('ep_session_number');

      final resumed = harness.build(now: () => base.add(const Duration(minutes: 10)));
      resumed.init();

      // Same device, so the session resumes and first_seen_at survives: only
      // the counter is repaired.
      expect(resumed.eventHeaders()['X-Session-Key'], firstKey);
      expect(harness.store.values['ep_first_seen_at'], firstSeen);
      expect(harness.store.values['ep_session_number'], '1');
      expect(harness.transport.identityBodies().last['session_number'], 1);
      resumed.dispose();
    });

    test('a negative counter is repaired, not carried, on every path', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      first.dispose();

      // -3 would become -2 through _rotate's `+= 1` and stay rejected, then
      // creep up one per launch forever.
      harness.store.values['ep_session_number'] = '-3';
      // Past the 30-minute window, so init takes the rotate branch — the one
      // the resume-only guard never covered.
      final rotated = harness.build(now: () => base.add(const Duration(hours: 2)));
      rotated.init();

      expect(harness.transport.identityBodies().last['session_number'], 1);
      expect(harness.store.values['ep_session_number'], '1');
      rotated.dispose();
    });

    test('metadata from another anonymous_id is never reported as this device', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      first.dispose();

      // A partial wipe or a restored backup: ep_aid is gone, the metadata is
      // not. SPEC §2 binds the two, so both reset rather than the new device
      // inheriting the old install's session count and first_seen_at.
      harness.store.values['ep_session_number'] = '47';
      harness.store.values['ep_first_seen_at'] = '2020-01-01T00:00:00.000Z';
      final priorKey = harness.store.values['ep_session_key'];
      harness.store.values.remove('ep_aid');

      final fresh = harness.build(now: () => base.add(const Duration(minutes: 5)));
      fresh.init();

      final body = harness.transport.identityBodies().last;
      expect(body['session_number'], 1);
      expect(body['first_seen_at'], isNot('2020-01-01T00:00:00.000Z'));
      // The session rotates with the device: identity_registry is keyed on
      // session_key alone, so one session_key must never carry two devices.
      expect(harness.store.values['ep_session_key'], isNot(priorKey));
      fresh.dispose();
    });

    test('metadata bound to a different anonymous_id is reset, not inherited', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      first.dispose();

      // ep_aid is present and intact, but the metadata records that it was
      // created under a different device — a restored backup, or a store
      // written by another install. Only the ep_meta_aid binding catches this;
      // `ep_aid == null` does not.
      harness.store.values['ep_meta_aid'] = '11111111-1111-4111-8111-111111111111';
      harness.store.values['ep_session_number'] = '47';
      harness.store.values['ep_first_seen_at'] = '2020-01-01T00:00:00.000Z';

      final fresh = harness.build(now: () => base.add(const Duration(minutes: 5)));
      fresh.init();

      final body = harness.transport.identityBodies().last;
      expect(body['session_number'], 1);
      expect(body['first_seen_at'], isNot('2020-01-01T00:00:00.000Z'));
      expect(harness.store.values['ep_meta_aid'], harness.store.values['ep_aid']);
      fresh.dispose();
    });

    test('an upgrade from a build without ep_meta_aid keeps its history', () {
      final base = DateTime.utc(2026, 7, 13);
      harness = Harness(now: () => base);

      final first = harness.build();
      first.init();
      first.dispose();
      final firstSeen = harness.store.values['ep_first_seen_at'];

      // A store written before ep_meta_aid existed. Reading its absence as
      // "belongs to another device" would reset first_seen_at and the session
      // count on every install's first launch after the SDK upgrade.
      harness.store.values.remove('ep_meta_aid');
      harness.store.values['ep_session_number'] = '9';

      final upgraded = harness.build(now: () => base.add(const Duration(minutes: 5)));
      upgraded.init();

      final body = harness.transport.identityBodies().last;
      expect(body['session_number'], 9);
      expect(body['first_seen_at'], firstSeen);
      // and the binding is backfilled, so the guard works from here on
      expect(harness.store.values['ep_meta_aid'], harness.store.values['ep_aid']);
      upgraded.dispose();
    });

    test('resumed lifecycle after >30 min rotates and re-registers', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        final firstKey = client.eventHeaders()['X-Session-Key'];

        async.elapse(const Duration(minutes: 31));
        client.handleLifecycle(AppLifecycleState.resumed);
        async.flushMicrotasks();

        expect(client.eventHeaders()['X-Session-Key'], isNot(firstKey));
        final identities = harness.transport.identityBodies();
        expect(identities, hasLength(2));
        expect(identities.last['session_number'], 2);
        client.dispose();
      });
    });
  });

  group('user transitions (SPEC §3)', () {
    test('setUser re-registers without rotating; clearUser rotates session only', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        final before = harness.transport.identityBodies().first;

        client.setUser('u-42');
        async.flushMicrotasks();
        final withUser = harness.transport.identityBodies().last;
        expect(withUser['user_id'], 'u-42');
        expect(withUser['anonymous_id'], before['anonymous_id']);
        expect(withUser['session_key'], before['session_key']);

        client.clearUser();
        async.flushMicrotasks();
        final cleared = harness.transport.identityBodies().last;
        expect(cleared.containsKey('user_id'), isFalse);
        expect(cleared['session_key'], isNot(before['session_key']));
        expect(cleared['session_number'], 2);
        expect(cleared['anonymous_id'], before['anonymous_id']);
        client.dispose();
      });
    });
  });

  group('engagement time (SPEC §4)', () {
    test('each event carries foreground ms since the previous event', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();

        async.elapse(const Duration(milliseconds: 1200));
        client.track('product_viewed');
        async.elapse(const Duration(milliseconds: 3400));
        client.track('product_viewed');
        client.flush();
        async.flushMicrotasks();

        final events = harness.transport.sentEvents();
        expect((events[0]['context']! as Map)['engagement_time_msec'], 1200);
        expect((events[1]['context']! as Map)['engagement_time_msec'], 3400);
        client.dispose();
      });
    });

    test('stopwatch pauses while backgrounded', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();

        async.elapse(const Duration(seconds: 2));
        client.handleLifecycle(AppLifecycleState.paused);
        async.elapse(const Duration(minutes: 5)); // background time not counted
        client.handleLifecycle(AppLifecycleState.resumed);
        async.elapse(const Duration(milliseconds: 500));
        client.track('product_viewed');
        client.flush();
        async.flushMicrotasks();

        final events = harness.transport.sentEvents();
        expect((events.last['context']! as Map)['engagement_time_msec'], 2500);
        client.dispose();
      });
    });
  });

  group('queue give-up (SPEC §7)', () {
    test('events older than 24h are dropped, not sent', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        harness.transport.manual = true;
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        harness.transport.completeNext(); // identity ok
        async.flushMicrotasks();

        client.track('product_viewed');
        client.flush();
        async.flushMicrotasks();
        harness.transport.completeNext(false); // send fails, stays queued
        async.flushMicrotasks();

        async.elapse(const Duration(hours: 25));
        final sendsBefore =
            harness.transport.calls.where((c) => c.$1 == '/v1/events').length;
        client.flush();
        async.flushMicrotasks();
        final sendsAfter =
            harness.transport.calls.where((c) => c.$1 == '/v1/events').length;
        expect(sendsAfter, sendsBefore); // nothing left to send
        client.dispose();
      });
    });
  });

  group('reportError (thin /v1/errors path)', () {
    test('posts immediately, bypassing the queue and the S4 gate', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 14);
        harness = Harness(now: () => base.add(async.elapsed));
        harness.transport.manual = true; // identity registration hangs
        final client = harness.build();
        client.init();
        async.flushMicrotasks();

        client.reportError(StateError('bad state'), StackTrace.current);
        async.flushMicrotasks();

        final errors =
            harness.transport.calls.where((c) => c.$1 == '/v1/errors').toList();
        expect(errors, hasLength(1));
        final body = errors.single.$2;
        expect(body['kind'], 'StateError');
        expect(body['message'], contains('bad state'));
        expect(body['stack'], isNotEmpty);
        expect(body['anonymous_id'], isNotNull);
        expect(body['session_key'], isNotNull);
        expect((body['sdk']! as Map)['name'], sdkName);
        expect(harness.transport.calls.where((c) => c.$1 == '/v1/events'), isEmpty);
        client.dispose();
      });
    });
  });

  group('eventHeaders (SPEC §8)', () {
    test('returns the tier-2 headers after init', () {
      harness = Harness(now: DateTime.now);
      final client = harness.build();
      client.init();
      final headers = client.eventHeaders('add_to_cart');
      expect(headers['X-Event'], 'add_to_cart');
      expect(headers['X-Session-Key'], isNotEmpty);
      expect(headers['X-Anonymous-Id'], isNotEmpty);
      client.dispose();
    });
  });

  group('setUserAttributes (SPEC §6.1)', () {
    test('is a no-op without a prior setUser call', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        final before = harness.transport.calls.length;

        client.setUserAttributes({'email': 'a@b.co'});
        async.flushMicrotasks();

        expect(harness.transport.calls.length, before); // no round-trip
        client.dispose();
      });
    });

    test('posts allowlisted attributes with user_id after setUser', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();

        client.setUser('u-42');
        async.flushMicrotasks();
        client.setUserAttributes({
          'first_name': 'Ali',
          'email': 'ali@example.com',
          'phone': '+9647701234567',
          'gender': 'male',
          'city': 'Baghdad',
        });
        async.flushMicrotasks();

        final identity = harness.transport.identityBodies().last;
        expect(identity['user_id'], 'u-42');
        final attributes = identity['attributes']! as Map;
        expect(attributes['first_name'], 'Ali');
        expect(attributes['email'], 'ali@example.com');
        expect(attributes['phone'], '+9647701234567');
        expect(attributes['gender'], 'male');
        expect(attributes['city'], 'Baghdad');
        client.dispose();
      });
    });

    test('drops keys outside the six-name allowlist locally', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        client.setUser('u-42');
        async.flushMicrotasks();

        client.setUserAttributes({
          'first_name': 'Ali',
          'ssn': '123-45-6789',        // not allowlisted
          'passport_number': 'X99',    // not allowlisted
        });
        async.flushMicrotasks();

        final attributes = harness.transport.identityBodies().last['attributes']! as Map;
        expect(attributes.keys, ['first_name']);
        expect(attributes.containsKey('ssn'), isFalse);
        expect(attributes.containsKey('passport_number'), isFalse);
        client.dispose();
      });
    });

    test('does not send when every provided key is dropped', () {
      fakeAsync((async) {
        final base = DateTime.utc(2026, 7, 13);
        harness = Harness(now: () => base.add(async.elapsed));
        final client = harness.build();
        client.init();
        async.flushMicrotasks();
        client.setUser('u-42');
        async.flushMicrotasks();
        final before = harness.transport.calls.length;

        client.setUserAttributes({'ssn': '123'});
        async.flushMicrotasks();

        expect(harness.transport.calls.length, before);
        client.dispose();
      });
    });
  });
}
